proposals/gateway-plusplus-phase1.md

```diff
@@ -0,0 +1,122 @@
+
+- Pinning Service API.
+    - 😡 Requires you to already be running a full IPFS node as a client.
+    - 😡 Is not transactional. You have to poll in order to tell when the data is actually available.
+    - 😃 Is intended to be multi-tenant and authenticated using a flexible bearer token.
+- HTTP RPC API
+    - 😡 Is single-tenant. No authentication system.
+        - The RPC API is designed for a single user to manage a remote node, it was not designed and built to be multi-tenant and would look very different if it were.
+    - 😡 Is not always transactional. Many write calls return before they are available.
+
+These interfaces were built with particular user profiles in mind. The users we’re seeing in NFTs (including our own service as a user of IPFS) are not well served by these interfaces and we need a new “project” definition (Gateway++) that is centered around these user needs.
+
+Developers are building applications that have to run in a lot of constrained environments that can’t support a full node. Strengthening the Gateway to serve these needs over HTTP leverages endpoints they are already using and provides the features they need in a stateless protocol that has ubiquitous support.
+
+We are never going to hear “Sorry, I can’t use HTTP in this environment” from a developer.
+
+## Phase 1
+
+We should expect this to be a long term project with additional proposals in due time.
+
+For now, we have a list of high priority tasks that need to be fixed.
+
+In [nft.storage](http://nft.storage) we have the following high priority needs:
+
+- Add the Pinning API to ipfs-cluster.
+- Add transactional CAR file uploads to the Pinning API.
+
+We also need to do the following to kick off the project.
+
+- Document configuring IPFS Gateway to pass Pinning API requests to a cluster or other IPFS node with the Pinning API enabled.
+- Configure and deploy our own gateway with this configuration. (We won’t be handing out auth tokens to anyone who doesn’t already have one, this is just to eat our own dogfood).
+
+#### What does done look like?
+
+Each work item in Phase 1 is something nft.storage needs so we have a builtin "first user" who would have all their requirements satisfied by this proposal.
+
+Once this is deployed you'll be able to write thin client interfaces with only a gateway URL and bearer token as configuration.
+
+####  What does success look like?
+
+Beyond just having satisfied the needs of nft.storage, this should also satisfy the needs that Cloudflare has and we would hopefully see support for IPFS in Cloudflare Workers as a thin
+client to their Gateway.
+
+#### Alternatives
+
+There are alternative approaches to building thin clients. The proposals around changing/improving the RPC API could be designed for this purpose,
```

Comment:
> regular IPFS node with disabled p2p transports and discovery and doing content via CAR import/export via Gateway

So what's left in a "regular IPFS node" when you strip out these bits? This description sounds just like what this proposal wants but without the notion of being a "regular IPFS node". But that probably comes back to the problems we have of "IPFS node" being something different for everyone! Has import via the gateway been something already on the table? How has that been imagined so far and is there an alternative here to pulling in the Pinning API to achieve this?

Symmetric use of CAR for import _and_ export would certainly be worth exploring as part of this proposal.

Reply:
>  So what's left in a "regular IPFS node" when you strip out these bits?

Integrity guaranteed provided by content addressing (data can be fetched in trustless manner) and ability to use IPLD for advanced data structures.

> Has import via the gateway been something already on the table? How has that been imagined so far and is there an alternative here to pulling in the Pinning API to achieve this?

Yes, we are planning to add DAG import/export directly to gateway endpoints (`/ipfs/`, `/ipns/`). Longer discussion in 
https://github.com/ipfs/in-web-browsers/issues/170 but tldr idea is:  
- Improve the concept of a writable gateway to support DAG import via `HTTP PUT /ipfs/{cid}`
- IPNS publishing could be as easy as `HTTP PUT /ipns/{libp2p-key}`

Reply:
How many of the environments that we're concerned about are unable to sustain a basic libp2p node that makes a single connection via TCP/WebSockets and really need to just have HTTP? As mentioned in some of my other comments we can efficiently use libp2p to transfer IPLD data between two peers as all the transports we support have bidirectional streaming, otherwise we lose efficiency by being unidirectional.

Reply:
> How many of the environments that we're concerned about are unable to sustain a basic libp2p node

I think it is less about resources and more about "deployment style", more specifically about preferring stateless-ness where possible.

A libp2p node is an active unit, requiring an actively running process, servicing of periodic protocol chatter, etc.

An HTTP client interface on the other hand is completely and utterly "dumb". You could drive such an "http-only ipfs-node" from a bash script, which is decidedly not possible today.

Reply:
A libp2p node could be setup to be effectively stateless though and easily spun up + down. The only reason the HTTP client appears stateless is because there's some library that under the hood is doing the TCP connection and managing it.

I can't speak to the state of js-ipfs/libp2p, but if I wanted to create a Go binary that operated as `ipfs-remote-pin <CAR file> <pinning-service-address> <auth-token>` that was easily usable from a bash script that would be a pretty small amount of work. This is with the introduction of no protocol modifications or additions.

Reply:
> ability to use IPLD for advanced data structures

CAR import gives us the entire write side of the IPLD interface. Existing block read interfaces are also sufficient (for now).

Reply:
> How many of the environments that we're concerned about are unable to sustain a basic libp2p node that makes a single connection via TCP/WebSockets and really need to just have HTTP?

Serverless (Lambda), Cloudflare Workers, and mobile devices. 

Pretty much all the highest growth application environments have trouble with long running processes and connections and prefer or require a stateless protocol.

Reply:
> have trouble with long running processes and connections and prefer or require a stateless protocol.

Is a long running HTTP upload exempt from this? If not then spinning up a temporary libp2p node shouldn't be very different.

Reply:
HTTP upload is not exempt, we can’t push too much data at once. We’re going to have to break up large files by encoding in the client and doing uploads under 100mb to get around CF Worker limits.

Reply:
Cloudflare Workers seems to have support for WebSockets https://blog.cloudflare.com/introducing-websockets-in-workers/ so using libp2p shouldn't be a problem there.

Reply:
It has an interface for being a websocket service but there’s no client in CF workers.